Add BitString.to_int and BitString.from_int

percontation opened the pull request:
These methods assist in converting between BitStrings and ints,
which is useful when bridging between Postgres and things like
bitmasks, enum.IntFlag, etc.

elprans commented on asyncpg/types.py:
```diff
@@ -228,6 +228,28 @@ def as_string(self):
+            raise ValueError("bitorder must be either 'little' or 'big'")
+
+    @classmethod
+    def from_int(cls, x, length, bitorder):
```
`length` should be an upper limit (like in `int.to_bytes`), not an actual bitstring length specification. The actual length can be unambiguously determined from the passed value.  I would also make it optional.

percontation replied:
I'm confused by this. The `length` argument of `int.to_bytes()` is not an upper limit, it's an actual byte length. And, as I understand it, it's not possible to unambiguously determine a bit length from a python integer, because there may be leading zeros (that aren't reflected in anything like `int.bit_length()`).

percontation replied:
Ah, okay. I didn't guess that Postgres would automatically convert shorter bits to longer bits. Will do.

percontation replied:
Okay, turns out length is necessary; asyncpg throws a StringDataLengthMismatchError instead of automatically padding with zeros. Which is just as well, cuz big-endian/little-endian/left-pad/right-pad is kinda confusing here :P

elprans commented on asyncpg/types.py:
```diff
@@ -228,6 +228,28 @@ def as_string(self):
 
         return s.strip()
 
+    def to_int(self, bitorder):
```
Is there a particular reason to make "bitorder" a thing?  At least, Postgres padding semantics seems to indicate that bitstrings have 'big' bitorder.  I would just go with this assumption.

percontation replied:
'little' is what I wanted for my case, for the typical little-endian advantage of giving consistent integer values for casts between Postgres bits of different lengths (and it gets even more annoying with varbits).

But yeah, I made the same assumption about Postgres padding / what people would expect from running `bin()` on the resulting integer, so I added bitorder by analogy to python's `int.to/from_bytes`. Potentially, unlike to/from_bytes, bitorder should have a default as 'big' here?

elprans replied:
> typical little-endian advantage of giving consistent integer values for casts between Postgres bits of different lengths

Can you elaborate on this please? It seems to me that "big" endians are actually more consistent given the right-padding (as done by `int.to_bytes(length=N)` and by `(CAST '111' AS bit(N))`

percontation replied:
With a little-endian bit-order, `B'101'::bit(10)` and `B'101'::bit(20)` result in the same integer. This is desirable if bits are being used to store collections of flags, because (for example) if a schema using bit(32) later needed to be migrated to bit(64) to support more flags, the values in Python would not need to change (assuming the usual C or Python enum.IntFlag conventions of assigning enum names to bits).

elprans replied:
If we go by the byte-order endianness definition where 'endian` actually means the *starting* bit, then 'little` is what we want here.

elprans replied:
Let's just assume the starting bit is least-significant, and omit the `bitorder` argument until someone comes along with a legitimate use case for it.

percontation replied:
Hmm. If we were going to only have one, it should be big-endian (despite the fact that that's not the one that personally want :cry:), because that's what Postgres currently uses when casting between ints and bits. Also, as you pointed out in the `from_int` case, inferring the length from the integer only makes sense when using big endian.

elprans replied:
Ah. I didn't realize the int casts were big-endian.  We can't have it nice here, it seems. Let's keep the argument then.

Please add docstrings to the new methods (with `.. versionadded:: 0.18.0`).

percontation commented:
(^ that latest commit doesn't reflect the pending "lets remove the bitorder option" discussion)

percontation commented:
I believe I addressed the comments now, lmk if there are still questions.

Also, added a `signed` kwarg to complete the analogy with int.to/from_bytes. Happy to remove that if it feels too superfluous. (I guess it would come in handy for people who want to read something like a signed int128_t in a SQL DB?)

elprans commented on asyncpg/types.py:
```diff
@@ -228,6 +228,93 @@ def as_string(self):
+        return x
+
+    @classmethod
+    def from_int(cls, x, length, bitorder=None, *, signed=True):
```
`signed` should be `False` by default.  Also, why not make bitorder default to `big` here as well?

percontation replied:
/shrug just a preference thing on both counts. I'll do it that way.

elprans commented:
Hi @percontation, can you squash the commits, please? I'd like to merge this.
